R/SCE-reducedDims.R

```diff
@@ -53,14 +53,59 @@ setReplaceMethod("reducedDimNames", c("SingleCellExperiment", "character"), func
 })
 
 #' @export
-setMethod("reducedDim", "SingleCellExperiment", function(x, type=1, withDimnames=TRUE) {
+setMethod("reducedDim", c("SingleCellExperiment", "missing"), function(x, type, withDimnames=TRUE) {
     x <- updateObject(x)
+    internals <- int_colData(x)[[.red_key]]
+
+    if (identical(ncol(internals), 0L)) {
+        stop(
+            "'reducedDim(<", class(x), ">, ...) ",
+            "length(reducedDims(<", class(x), ">)) is 0'")
+    }
 
+    reducedDim(x, 1L, withDimnames)
+})
+
+#' @export
+setMethod("reducedDim", c("SingleCellExperiment", "numeric"), function(x, type=1, withDimnames=TRUE) {
+    x <- updateObject(x)
     internals <- int_colData(x)[[.red_key]]
-    out <- internals[,type]
-    if (!is.null(out) && withDimnames) {
+
+    out <- tryCatch({
+        internals[, type]
+    }, error=function(err) {
+        stop("'reducedDim(<", class(x), ">, type=\"numeric\", ...)' ",
+             "invalid subscript 'type'\n", conditionMessage(err))
+    })
+
+    if (withDimnames) {
         rownames(out) <- colnames(x)
     }
+
+    out
+})
+
+#' @export
+setMethod("reducedDim", c("SingleCellExperiment", "character"), function(x, type, withDimnames=TRUE) {
+    msg <- paste0("'reducedDim(<", class(x), ">, type=\"character\", ...)' ",
+                  "invalid subscript 'type'")
+
+    x <- updateObject(x)
+    internals <- int_colData(x)[[.red_key]]
+
+    out <- tryCatch({
+        internals[, type]
+    }, error=function(err) {
+        NULL
+    })
+
+    if (is.null(out))
```

Comment:
This should go into the body of the `tryCatch`, don't know why you need to catch the `NULL` here.

Reply:
Oh wait, just replace this with the dep warning. But make sure you use `if (...) {`.